Modules/_testexternalinspection.c

```diff
@@ -431,44 +456,604 @@ read_string(pid_t pid, _Py_DebugOffsets* debug_offsets, void* address, char* buf
+    struct _Py_DebugOffsets* offsets,
+    struct _Py_AsyncioModuleDebugOffsets* async_offsets,
+    uintptr_t task_address,
+    PyObject *render_to
+) {
+    char is_task;
+    int err = read_char(
+        pid,
+        task_address + async_offsets->asyncio_task_object.task_is_task,
+        &is_task);
+    if (err) {
+        return -1;
+    }
+
+    uintptr_t refcnt;
+    read_ptr(pid, task_address + sizeof(Py_ssize_t), &refcnt);
+
+    PyObject* result = PyList_New(0);
+    if (result == NULL) {
         return -1;
     }
 
-    char function_name[256];
-    if (read_string(pid, offsets, address_of_function_name, function_name, sizeof(function_name)) != 0) {
+    PyObject *call_stack = PyList_New(0);
+    if (call_stack == NULL) {
         return -1;
     }
+    if (PyList_Append(result, call_stack)) {
+        Py_DECREF(call_stack);
+        goto err;
+    }
+    /* we can operate on a borrowed one to simplify cleanup */
+    Py_DECREF(call_stack);
+
+    if (is_task) {
+        PyObject *tn = parse_task_name(pid, offsets, async_offsets, task_address);
+        if (tn == NULL) {
+            goto err;
+        }
+        if (PyList_Append(result, tn)) {
+            Py_DECREF(tn);
+            goto err;
+        }
+        Py_DECREF(tn);
+
+        uintptr_t coro_addr;
+        err = read_py_ptr(
+            pid,
+            task_address + async_offsets->asyncio_task_object.task_coro,
+            &coro_addr);
+        if (err) {
+            goto err;
+        }
+
+        if ((void*)coro_addr != NULL) {
+            err = parse_coro_chain(
+                pid,
+                offsets,
+                async_offsets,
+                coro_addr,
+                call_stack
+            );
+            if (err) {
+                goto err;
+            }
+
+            if (PyList_Reverse(call_stack)) {
+                goto err;
+            }
+        }
+    }
 
-    PyObject* py_function_name = PyUnicode_FromString(function_name);
+    if (PyList_Append(render_to, result)) {
+        goto err;
+    }
+
+    PyObject *awaited_by = PyList_New(0);
+    if (awaited_by == NULL) {
+        goto err;
+    }
+    if (PyList_Append(result, awaited_by)) {
+        Py_DECREF(awaited_by);
+        goto err;
+    }
+    /* we can operate on a borrowed one to simplify cleanup */
+    Py_DECREF(awaited_by);
+
+    if (parse_task_awaited_by(pid, offsets, async_offsets, task_address, awaited_by)) {
+        goto err;
+    }
+
+    return 0;
+
+err:
+    Py_DECREF(result);
+    return -1;
+}
+
+static int
+parse_tasks_in_set(
+    int pid,
+    struct _Py_DebugOffsets* offsets,
+    struct _Py_AsyncioModuleDebugOffsets* async_offsets,
+    uintptr_t set_addr,
+    PyObject *awaited_by
+) {
+    uintptr_t set_obj;
+    if (read_py_ptr(
+            pid,
+            set_addr,
+            &set_obj)
+    ) {
+        return -1;
+    }
+
+    Py_ssize_t set_len;
+    if (read_ssize_t(
+            pid,
+            set_obj + offsets->set_object.used,
```

Comment:
@pablogsal we might want to use both `used` and `mask` here just in case.

Reply:
Sup, I am recovering from the jet lag but that's my next update

Reply:
fixed